clients/src/main/java/org/apache/kafka/common/record/MemoryRecords.java

```diff
@@ -112,7 +112,10 @@ public void append(long offset, byte[] key, byte[] value) {
      * to accept this single record.
      */
     public boolean hasRoomFor(byte[] key, byte[] value) {
-        return this.writable && this.compressor.numRecordsWritten() == 0 ?
+        if (!this.writable)
+            return false;
+
+        return this.compressor.numRecordsWritten() == 0 ?
```

Comment:
Definitely not the most readable code. 
But if that's the actual upstream patch, I would keep like this to avoid drift.